inode.c

```diff
@@ -609,35 +748,70 @@ static int simplefs_link(struct dentry *old_dentry,
     struct inode *inode = d_inode(old_dentry);
     struct super_block *sb = inode->i_sb;
     struct simplefs_inode_info *ci_dir = SIMPLEFS_INODE(dir);
-    struct simplefs_dir_block *dir_block;
-    struct buffer_head *bh;
-    int f_pos = -1, ret = 0, i = 0;
+    struct simplefs_file_ei_block *eblock = NULL;
+    struct simplefs_dir_block *dblock;
+    struct buffer_head *bh = NULL, *bh2 = NULL;
+    int ret = 0, alloc = false, bno = 0;
+    int ei = 0, bi = 0, fi = 0;
 
-    bh = sb_bread(sb, ci_dir->dir_block);
+    bh = sb_bread(sb, ci_dir->ei_block);
     if (!bh)
         return -EIO;
-    dir_block = (struct simplefs_dir_block *) bh->b_data;
+    eblock = (struct simplefs_file_ei_block *) bh->b_data;
 
-    if (dir_block->files[SIMPLEFS_MAX_SUBFILES - 1].inode != 0) {
+    if (eblock->nr_files == SIMPLEFS_MAX_SUBFILES) {
         ret = -EMLINK;
         printk(KERN_INFO "directory is full");
         goto end;
     }
 
-    for (i = 0; i < SIMPLEFS_MAX_SUBFILES; i++) {
-        if (dir_block->files[i].inode == 0) {
-            f_pos = i;
-            break;
+    ei = eblock->nr_files / SIMPLEFS_FILES_PER_EXT;
+    bi = eblock->nr_files % SIMPLEFS_FILES_PER_EXT
+         / SIMPLEFS_FILES_PER_BLOCK;
+    fi = eblock->nr_files % SIMPLEFS_FILES_PER_EXT
```

Comment:
Can you avoid the computation by re-using the value of `bi` for `fi`?

Reply:
I'm not quite understand what you mean, I guess you wish to reduce the usage of modulo.
Do you mean replacing `fi` by 
`fi = eblock->nr_files - ei * SIMPLEFS_FILES_PER_EXT - bi * SIMPLEFS_FILES_PER_BLOCK`?

Reply:
What I meant was to shorten the expression.

Reply:
I found that `fi` could be changed to `fi = eblock->nr_files % SIMPLEFS_FILES_PER_BLOCK`